### PR TITLE
Prepend the web dir in `FigureBuilder::fromUrl()`

### PR DESCRIPTION
**Pull request by fritzmg**
Currently you cannot use the `FigureBuilder` to process images that only reside in the `public/` directory. e.g. images that you have stored there manually (as they are public assets and do not need to be managed by the file manager) or may be they are automatically generated (e.g. via Webpack etc.).

i.e. the following does not work:

```
public/
    foobar/
        image.jpg
```
```php
$figureBuilder
    ->fromUrl('foobar/image.jpg')
    ->build()
;
```

This is unexpected as `fromUrl` would imply that any file that is publicly accessible via an URL should work. But currently `fromUrl` only works for images that are symlinked from root to `public`.

This PR automatically prepends the `%contao.web_dir%` within `FigureBuilder::fromUrl` - except for URLs that start with `files/` or `assets/` as these are symlinked folders from the root directory. We must not prepend the web dir to `files/` so that `fromPath` automatically loads the meta data from `tl_files`.

This will also make it possible to process images from an asset URL:

```php
$figureBuilder
    ->fromUrl($this->packages->getUrl('images/app_image.jpg'))
    ->build()
;
```

_Note:_ Previously the tests used `<root>/images/dummy.jpg` for example for the `fromUrl` tests, which was not a realistic test, as such files would not have been accessible via a URL. Thus these tests had to be changed now.

### CODE REVIEW

File: core-bundle/src/Image/Studio/FigureBuilder.php
Comment (ausi) [inline core-bundle/src/Image/Studio/FigureBuilder.php]: Do we need an exception for `/system/themes` or just `/system` too?

Comment (fritzmg) [reply core-bundle/src/Image/Studio/FigureBuilder.php]: I don't think so. Technically we also don't need it for `assets/`. The only case where it is important is `files/` as otherwise the meta data would not get loaded. But that does not matter for any other folder.

**Review by ausi — COMMENTED** [review]
✅ Approved. But we should wait for @m-vo s approval IMO.

**Comment by fritzmg** [comment]
> But we should wait for @m-vo s approval IMO.

Yes. Also I need to fix the tests anyway :)

**Comment by leofeyer** [comment]
> We must not prepend the web dir to `files/` so that `fromPath` automatically loads the meta data from `tl_files`.

It seems a little weird to me that we have to prepend to some paths but must not prepend to others in order not to break yet another logic. Can we standardize this to fit all cases?

**Comment by fritzmg** [comment]
We could instead change `fromPath` so that it checks for `public/files/` and automatically removes `public/` when fetching the file from the database. But not sure that really improves anything?

**Comment by ausi** [comment]
>> Can we standardize this to fit all cases?
>
> We could instead change `fromPath` so that it checks for `public/files/` and automatically removes `public/` when fetching the file from the database.

Yes I think this would be the right way. This way everyone can use `fromPath()` with files from `public/` too. So this would make both `fromUrl()` and `fromPath()` more robust.

**Comment by fritzmg** [comment]
Changed in ab35e8d. Now the "auto detect dbafs path" check checks for `…/files` and `…/public/files`. Thus in all cases the meta data will be loaded from the DBAFS:

```php
$figureBuilder->fromPath('public/files/foobar/image.jpg');
$figureBuilder->fromPath('files/foobar/image.jpg');
$figureBuilder->fromUrl('files/foobar/image.jpg'); // = …/public/files/foobar/image.jpg
```

And these work seamlessly as well:

```php
$figureBuilder->fromPath('public/assets/contao/images/facebook.svg');
$figureBuilder->fromPath('assets/contao/images/facebook.svg');
$figureBuilder->fromUrl('assets/contao/images/facebook.svg');
```

File: core-bundle/src/Image/Studio/FigureBuilder.php
Comment (leofeyer) [inline core-bundle/src/Image/Studio/FigureBuilder.php:299]: No more `ltrim()`?

Comment (fritzmg) [reply core-bundle/src/Image/Studio/FigureBuilder.php:299]: `Path::join` already takes care of that.

File: core-bundle/src/Image/Studio/FigureBuilder.php
Comment (m-vo) [inline core-bundle/src/Image/Studio/FigureBuilder.php]: How about using a function so that things are only evaluated when needed?

```suggestion
        $getDbafsPath = static function(): ?string {
            if($Path::isBasePath(Path::join($this->webDir, $this->uploadPath), $path)) {
                return Path::makeRelative($path, $this->webDir);
            }
    
            if(Path::isBasePath(Path::join($this->projectDir, $this->uploadPath)) {
                return $path;
            }
               
            return null;
        }

        if ($autoDetectDbafsPaths && null !== ($dbafsPath = $getDbafsPath())) {
            $filesModel = $this->getFilesModelAdapter()->findByPath($dbafsPath);
```

**Review by m-vo — DISMISSED** [review]
Makes sense to me, thanks for the work! :+1: :slightly_smiling_face:

**Review by ausi — DISMISSED** [review]

**Review by m-vo — APPROVED** [review]

**Review by ausi — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @fritzmg.

**Pull request merged**
